feat: Use postgres as a queue

uhthomas opened the pull request:
We've been keen to try this for a while as it means we can remove redis as a dependency, which makes Immich easier to setup and run.

This replaces bullmq with a bespoke postgres queue. Jobs in the queue are processed either immediately via triggers and notifications, or eventually if a notification is missed.

## Description

<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Fixes # (issue)

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

This change has mostly been manually tested both locally and with the preview instances. We've tested that jobs run correctly, pause/resume, clear, etc. We've also done some benchmarks and compared results against bullmq (todo, share results).

<details><summary><h2>Screenshots (if appropriate)</h2></summary>

<!-- Images go below this line. -->

</details>

<!-- API endpoint changes (if relevant)
## API Changes
The `/api/something` endpoint is now `/api/something-else`
-->

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

uhthomas commented on .vscode/settings.json:
yeah, I'll revert this

tlvince commented:
Postgres was [previously used](https://github.com/immich-app/immich/pull/10801) for IPC, which lead to frequent disk writes. Does this change reintroduce and/or address this?

uhthomas commented:
@tlvince Thanks for your interest! The queue uses NOTIFY/LISTEN and so yes this will prevent disks from idling and will mean the database is more active. We have discussed this a believe it may be a worthwhile trade-off. Similar behaviour should have been observed if persistence was enabled for Redis too.

The documentation does make note that the database should be installed on an SSD, and not on a pool of HDDs or network storage. If this suggestion is followed, then idling and unnecessary disk wear should not be a concern.

https://immich.app/docs/install/requirements

tlvince commented:
Yeah, I've been running a few single-server deployments with Redis persistence disabled, which anecdotally has been fine. It'd be nice if in-memory queues remained possible.

uhthomas commented:
FWIW @tlvince, we don't think using postgres as queue should be a regression on this front. We will take a look at improving the socket.io adapter so that disk idling is possible :crossed_fingers:

tlvince commented:
Thanks for looking into it. Last time I looked, socket.io's heartbeat triggers the Postgres commit which is needed for `notify` to work. There're some remnants of the findings in https://github.com/immich-app/immich/discussions/8009

uhthomas committed:
feat: Use postgres as a queue

We've been keen to try this for a while as it means we can remove redis as a
dependency, which makes Immich easier to setup and run.

This replaces bullmq with a bespoke postgres queue. Jobs in the queue are
processed either immediately via triggers and notifications, or eventually if a
notification is missed.

NicholasFlamy commented:
> FWIW @tlvince, we don't think using postgres as queue should be a regression on this front. We will take a look at improving the socket.io adapter so that disk idling is possible 🤞

That'd be great because I'm also interested in this.

jrasm91 commented:
This is unlikely to get merged so I will go ahead and close it for now.
